### PR TITLE
[pickers] Reset input value when locale is modified

### LINKED ISSUE #5243: [LocalizationProvider] changing locale doesn't update DatePicker content
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

Current behavior is visible in this example : [https://mui.com/x/react-date-pickers/localization/#date-engine-locale](https://mui.com/x/react-date-pickers/localization/#date-engine-locale)

- Open this codesandbox.
- Change DatePicker component by adding this prop `views={["year", "month"]}`
- Refresh CodeSandbox
- Swap language to fr
- content isn't updated, and remains in russian

image

### Expected behavior 🤔

after swaping locale,
i would expect to see 
image

### Steps to reproduce 🕹

Explained above

### Context 🔦

if you display a form with initial data, and then switch from a language to another, DatePicker content remains the same.
Which, I think, should not be the case.

### Your environment 🌎

Browser : Chrome  102.0.5005.115

<details>
  <summary><code>npx @mui/envinfo</code></summary>

 System:
    OS: Linux 4.19 Ubuntu 18.04.6 LTS (Bionic Beaver)
  Binaries:
    Node: 12.22.12 - /usr/bin/node
    Yarn: Not Found
    npm: 8.3.1 - ~/.npm-global/bin/npm
  Browsers:
    Chrome: Not Found
    Firefox: Not Found
  npmPackages:
    @emotion/react: 11.9.3 => 11.9.3 
    @emotion/styled: 11.9.3 => 11.9.3 
    @mui/base:  5.0.0-alpha.85 
    @mui/icons-material: 5.8.4 => 5.8.4 
    @mui/material: 5.8.4 => 5.8.4 
    @mui/private-theming:  5.8.4 
    @mui/styled-engine:  5.8.0 
    @mui/system:  5.8.4 
    @mui/types:  7.1.4 
    @mui/utils:  5.8.4 
    @mui/x-data-grid: 5.12.1 => 5.12.1 
    @mui/x-date-pickers: 5.0.0-alpha.6 => 5.0.0-alpha.6 
    @types/react: 17.0.39 => 17.0.39 
    react: 17.0.2 => 17.0.2 
    react-dom: 17.0.2 => 17.0.2 
    typescript: 4.7.3 => 4.7.3 
</details>

### Order ID 💳 (optional)

_No response_

### LINKED ISSUE #5305: [DatePicker] [LocalizationProvider] Date formatting incorrect after switching locale
### Duplicates

- [X] I have searched the existing issues

### Latest version

- [X] I have tested the latest version

### Current behavior 😯

When changing locales, the formatting of the selected date follows a mix of the previously selected and newly selected locales' rules. Selecting a different date via the picker fixes this until the locale changes again.

Example: `ru` -> `en-US` results in dates being formatted with the day first, month second.

### Expected behavior 🤔

When changing locales, the formatting of the selected date follows the newly selected locales' rules.

Example: `ru` -> `en-US` results in dates being formatted with the month first, day second.

### Steps to reproduce 🕹

Steps:

1. View the [Localization page in the docs](https://mui.com/x/react-date-pickers/localization/#date-engine-locale)
   - `RU` is selected by default
   - `23.06.2022` ✅
3. Select `EN` to switch the locale
   - `23/06/2022` ❌
4. Select a different date
   - `06/24/2022` ✅
5. Select `RU` to switch the locale
   - `06.24.2022` ❌

### Context 🔦

We are asynchronously importing locales from date-fns which means LocalizationProvider will initially be rendered with no locale specified and shortly after, the imported locale will be set.

### Your environment 🌎

The environment is mui.com. I can reproduce the issue on MacOS using Edge, Chrome, Firefox, and Safari.

### Order ID 💳 (optional)

_No response_

### PR DESCRIPTION
**Pull request by alexfauquette**
Follow up of #5265 (merge/improve before)

Fix #5305
Fix #5243

### CODE REVIEW

File: packages/x-date-pickers/src/internals/hooks/useMaskedInput.tsx [hunk 1]
```diff
@@ -80,19 +80,22 @@ export const useMaskedInput = <TInputDate, TDate>({
...
+  const prevLocale = React.useRef<Locale | string>(utils.locale);
 
   React.useEffect(() => {
     const rawValueChange = rawValue !== prevRawValue.current;
+    const localeChange = utils.locale !== prevLocale.current;
     prevRawValue.current = rawValue;
+    prevLocale.current = utils.locale;
 
-    if (!rawValueChange) {
+    if (!rawValueChange && !localeChange) {
```
Comment (LukasTy) [inline packages/x-date-pickers/src/internals/hooks/useMaskedInput.tsx:91]: Nitpick: IMHO, `isLocaleChanged` or `localeChanged` variable name would better suite a boolean field.  
The same would apply to the `rawValueChange` field. 😃

**Review by LukasTy — APPROVED** [review]
LGTM. 👍

File: packages/x-date-pickers/src/internals/hooks/useMaskedInput.tsx (hunk 1 above)
Comment (flaviendelangle) [reply packages/x-date-pickers/src/internals/hooks/useMaskedInput.tsx:91]: Agree, I would go with `hasLocaleChanged` here

Comment (alexfauquette) [reply packages/x-date-pickers/src/internals/hooks/useMaskedInput.tsx]: I moved to `xxxHasChanged`

Comment (LukasTy) [reply packages/x-date-pickers/src/internals/hooks/useMaskedInput.tsx]: Interesting change, but IMO better than the original anyways. 😉

**Pull request merged**